tokio/src/runtime/thread_pool/queue.rs

```diff
@@ -0,0 +1,550 @@
+
+                self.inner.buffer[idx].with_mut(|ptr| {
+                    // Write the task to the slot
+                    unsafe {
+                        ptr::write((*ptr).as_mut_ptr(), task);
+                    }
+                });
+
+                // Make the task available
+                self.inner.tail.store(tail.wrapping_add(1), Release);
+
+                return;
+            }
+
+            // The local buffer is full. Push a batch of work to the inject
+            // queue.
+            match self.push_overflow(task, head, tail, inject) {
+                Ok(_) => return,
+                // Lost the race, try again
+                Err(v) => task = v,
+            }
+
+            atomic::spin_loop_hint();
+        }
+    }
+
+    /// Moves a batch of tasks into the inject queue.
+    ///
+    /// This will temporarily make some of the tasks unavailable to stealers.
+    /// Once `push_overflow` is done, a notification is sent out, so if other
+    /// workers "missed" some of the tasks during a steal, they will get
+    /// another opportunity.
+    #[inline(never)]
+    fn push_overflow(
+        &mut self,
+        task: task::Notified<T>,
+        head: u32,
+        tail: u32,
+        inject: &Inject<T>,
+    ) -> Result<(), task::Notified<T>> {
+        const BATCH_LEN: usize = LOCAL_QUEUE_CAPACITY / 2 + 1;
+
+        let n = tail.wrapping_sub(head) / 2;
+        assert_eq!(n as usize, LOCAL_QUEUE_CAPACITY / 2, "queue is not full");
+
+        // Claim a bunch of tasks
+        //
+        // We are claiming the tasks **before** reading them out of the buffer.
+        // This is safe because only the **current** thread is able to push new
+        // tasks.
+        //
+        // There isn't really any need for memory ordering... Relaxed would
+        // work. This is because all tasks are pushed into the queue from the
+        // current thread (or memory has been acquired if the local queue handle
+        // moved).
+        let actual = self.inner.head.compare_and_swap(head, head + n, Release);
+        if actual != head {
+            // We failed to claim the tasks, losing the race. Return out of
+            // this function and try the full `push` routine again. The queue
+            // may not be full anymore.
+            return Err(task);
+        }
+
+        // link the tasks
+        for i in 0..n {
+            let j = i + 1;
+
+            let i_idx = (i + head) as usize & MASK;
+            let j_idx = (j + head) as usize & MASK;
+
+            // Get the next pointer
+            let next = if j == n {
+                // The last task in the local queue being moved
+                task.header().into()
+            } else {
+                self.inner.buffer[j_idx].with(|ptr| unsafe {
+                    let value = (*ptr).as_ptr();
+                    (*value).header().into()
+                })
+            };
+
+            self.inner.buffer[i_idx].with_mut(|ptr| unsafe {
```

Comment:
Safety here (and above) relies on the CAS. The unsafe just above this one is only okay because we special-case `j == n`.